### PR TITLE
Update PercentEscaper to Support UTF-8

### LINKED ISSUE #10216: [QUERY]Error occurs when downloading  files with Japanese filename from Azure Files with Azure SDK for Java on Eclipse
Title: Error occurs when downloading  files with Japanese filename from Azure Files with Azure SDK for Java on Eclipse

---Wanted answer---
Does SDK support downloading files with Japanese filename? 
Is there any workaround ?
Is any plans to fix this(support download a file with Japanese file name )?

---replication procedure---
1. Upload Japanese language file to Azure Files.
file name: 日本語.txt

2. Configure Eclipse. Install Java and Azure SDK.
azure-storage-file-share 12.2.0
azure-storage-common 12.4.0
azure-core 1.2.0
azure-core-http-netty 1.3.0

3.Execute code(Error occur by fileClient.openInputStream();)

/**
*
*/
package TestAzureFiles;
import com.azure.core.http.HttpClient;
import com.azure.storage.file.share.ShareClient;
import com.azure.storage.file.share.ShareFileClient;
import com.azure.storage.file.share.ShareServiceClient;
import com.azure.storage.file.share.ShareServiceClientBuilder;
/**
* @author ...
*
*/
public class TestModule001 {

    public static final String CONNECTION_STRING =
            "DefaultEndpointsProtocol=https;AccountName=<ACCOUNT_NAME>;AccountKey=<ACCOUNT_KEY>;EndpointSuffix=core.windows.net";
   /**
     * @param args
     */
    public static void main(String[] args)  {
        HttpClient httpClient1 = HttpClient.createDefault();
        ShareServiceClient serviceClient = new ShareServiceClientBuilder().connectionString(CONNECTION_STRING)
        .httpClient(httpClient1).buildClient();
        String id = "120031026002037";
        // Get File Sharing Client
        ShareClient shareClient = serviceClient.getShareClient(id);
        // What I tried
        String filepath = "日本語.txt";
        ShareFileClient fileClient = shareClient.getFileClient((String) filepath);
        fileClient.getFilePath();
        fileClient.openInputStream();
    }
}

---logs of error(java.lang.Exception)---
Exception in thread "main" java.lang.ArrayIndexOutOfBoundsException: Index 26085 out of bounds for length 256
    at com.azure.core.http.rest.PercentEscaper.escape(PercentEscaper.java:93)
    at com.azure.core.http.rest.SwaggerMethodParser.applySubstitutions(SwaggerMethodParser.java:566)
    at com.azure.core.http.rest.SwaggerMethodParser.setPath(SwaggerMethodParser.java:274)
    at com.azure.core.http.rest.RestProxy.createHttpRequest(RestProxy.java:218)
    at com.azure.core.http.rest.RestProxy.invoke(RestProxy.java:131)
    at com.azure.storage.file.share.implementation.$Proxy7.getProperties(Unknown Source)
    at com.azure.storage.file.share.implementation.FilesImpl.getPropertiesWithRestResponseAsync(FilesImpl.java:317)
    at com.azure.storage.file.share.ShareFileAsyncClient.getPropertiesWithResponse(ShareFileAsyncClient.java:880)
    at com.azure.storage.file.share.ShareFileAsyncClient.lambda$getPropertiesWithResponse$22(ShareFileAsyncClient.java:870)
    at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:118)
    at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:121)
    at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2274)
    at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
    at reactor.core.publisher.MonoFlatMap$FlatMapMain.onSubscribe(MonoFlatMap.java:103)
    at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:90)
    at reactor.core.publisher.MonoCurrentContext.subscribe(MonoCurrentContext.java:35)
    at reactor.core.publisher.Mono.subscribe(Mono.java:4210)
    at reactor.core.publisher.Mono.block(Mono.java:1665)
    at com.azure.storage.file.share.StorageFileInputStream.<init>(StorageFileInputStream.java:49)
    at com.azure.storage.file.share.ShareFileClient.openInputStream(ShareFileClient.java:111)
    at com.azure.storage.file.share.ShareFileClient.openInputStream(ShareFileClient.java:99)
    at TestAzureFiles.TestModule001.main(TestModule001.java:42)
    Suppressed: java.lang.Exception: #block terminated with an error
        at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:99)
        at reactor.core.publisher.Mono.block(Mono.java:1666)
        ... 4 more

---cause of error---
Error occurs when SDK for Java creates download path. Perhaps an error in percent encoding.
The conversion character is defined as 256 characters.
In other cases such as Japanese, will occur "ArrayIndexOutOfBoundsException".

Problem in the Azure-SDK-for-Java processing of " original.charAt(i)".
Please confirm in the source code line 81.

URL: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PercentEscaper.java#L81
(line 81: char c = original.charAt(i);)

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #10216 

This PR updates `PercentEncoder` to support encoding UTF-8 characters.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PercentEscaper.java [hunk 1]
```diff
@@ -3,104 +3,225 @@
...
-     * @param usePlusForSpace escape ' ' as '+' if true, "%20" otherwise
+     *
+     * @param safeCharacters Collection of characters that won't be escaped.
+     * @param usePlusForSpace If true {@code ' '} will be escaped as {@code '+'} instead of {@code "%20"}.
      */
-    PercentEscaper(String safeChars, boolean usePlusForSpace) {
-        for (int i = 0; i != safeChars.length(); i++) {
-            this.safeChars.add(safeChars.charAt(i));
-        }
+    PercentEscaper(String safeCharacters, boolean usePlusForSpace) {
```
Comment (srnagar) [inline sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PercentEscaper.java:31]: We should verify that space is not in `safeCharacters` if  `usePlusForSpace` is true.

File: sdk/core/azure-core/src/test/java/com/azure/core/http/rest/PercentEscaperTests.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            Arguments.arguments(defaultEscaper, "¢", "%C2%A2"),
+            Arguments.arguments(defaultEscaper, "ह", "%E0%A4%B9"),
+            Arguments.arguments(defaultEscaper, "€", "%E2%82%AC"),
+            Arguments.arguments(defaultEscaper, "한", "%ED%95%9C"),
+            Arguments.arguments(defaultEscaper, "円", "%E5%86%86"),
+            Arguments.arguments(defaultEscaper, "\uD800\uDF48", "%F0%90%8D%88"),
+            Arguments.arguments(defaultEscaper, "日本語", "%E6%97%A5%E6%9C%AC%E8%AA%9E"),
+            Arguments.arguments(defaultEscaper, " ", "%20"),
+            Arguments.arguments(new PercentEscaper(null, true), " ", "+"),
+            Arguments.arguments(new PercentEscaper("$", false), "$", "$")
```
Comment (srnagar) [inline sdk/core/azure-core/src/test/java/com/azure/core/http/rest/PercentEscaperTests.java:38]: These would be a good test cases to include as well.
`Arguments.arguments(new PercentEscaper( "ह", false),  "ह", "ह")`

`new PercentEscaper(" ", true); // this should throw exception`

File: sdk/core/azure-core/src/test/java/com/azure/core/http/rest/SwaggerMethodParserTests.java [hunk 3]
```diff
@@ -331,7 +331,7 @@ public void querySubstitution(Method method, Object[] arguments, Map<String, Str
             Arguments.of(substitution, null, null),
             Arguments.of(substitution, toObjectArray("raw", true), createExpectedParameters("raw", true)),
             Arguments.of(substitution, toObjectArray(null, true), createExpectedParameters(null, true)),
-            Arguments.of(substitution, toObjectArray("{sub1}", false), createExpectedParameters("%7bsub1%7d", false)),
+            Arguments.of(substitution, toObjectArray("{sub1}", false), createExpectedParameters("%7Bsub1%7D", false)),
```
Comment (srnagar) [inline sdk/core/azure-core/src/test/java/com/azure/core/http/rest/SwaggerMethodParserTests.java:334]: This is a semantic breaking change for users as it now encodes strings differently and requires minor version bump.

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PercentEscaper.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PercentEscaper.java:31]: Good idea, will add a check for that.

File: sdk/core/azure-core/src/test/java/com/azure/core/http/rest/PercentEscaperTests.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/test/java/com/azure/core/http/rest/PercentEscaperTests.java:38]: PercentEscaper is package private with public static instance available, shouldn't see these inputs, can add tests though.

File: sdk/core/azure-core/src/test/java/com/azure/core/http/rest/SwaggerMethodParserTests.java (hunk 3 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/test/java/com/azure/core/http/rest/SwaggerMethodParserTests.java:334]: I believe hex characters should be invariant on casing, I just copied the hex array creation from another of our classes and it used uppercase.

File: sdk/core/azure-core/src/test/java/com/azure/core/http/rest/PercentEscaperTests.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/test/java/com/azure/core/http/rest/PercentEscaperTests.java:38]: Added

File: sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PercentEscaper.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core/src/main/java/com/azure/core/http/rest/PercentEscaper.java:31]: Added

**Pull request merged**